File: Compose-Amphibians-App/app/src/main/java/com/example/myapplication/data/AppContainer.kt
package com.example.myapplication.data

import com.example.myapplication.network.AmphibiansApiService
import com.jakewharton.retrofit2.converter.kotlinx.serialization.asConverterFactory
import kotlinx.coroutines.Dispatchers
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import retrofit2.Retrofit

interface AppContainer {
    val amphibiansInfoRepository: AmphibiansInfoRepository
}


class DefaultAppContainer : AppContainer {

    private val baseUrl = "https://android-kotlin-fun-mars-server.appspot.com/"

    private val retrofit: Retrofit = Retrofit.Builder()
        .addConverterFactory(Json.asConverterFactory("application/json".toMediaType()))
        .baseUrl(baseUrl)
        .build()

    private val amphibiansApi: AmphibiansApi by lazy {
        retrofit.create(AmphibiansApiService::class.java)
    }

    private val amphibiansRemoteDataSource: AmphibiansRemoteDataSource by lazy {
        AmphibiansRemoteDataSource(
            amphibiansApi,
            Dispatchers.IO
        )
    }

    override val amphibiansInfoRepository: AmphibiansInfoRepository by lazy {
        NetworkAmphibiansInfoRepository(amphibiansRemoteDataSource)
    }
}

File: SwiftUI-Amphibians-App/Amphibians App/Data/AppContainer.swift
//
//  AppContainer.swift
//  Amphibians App
//
//  Created by Jaehwa Noh on 1/8/24.
//

protocol AppContainer {
    var amphibiansInfoRepository: AmphibiansInfoRepository { get }
}

class DefaultAppContainer: AppContainer {
    
    let baseURL = "https://android-kotlin-fun-mars-server.appspot.com/"
    
    lazy var amphibiansApi = AmphibiansApiService(baseURL: baseURL)
    
    lazy var amphibiansRemoteDataSource = AmphibiansRemoteDataSource(
        amphibiansApi: amphibiansApi,
        ioTask: .userInitiated
    )
    
    var amphibiansInfoRepository: AmphibiansInfoRepository {
        NetworkAmphibiansInfoRepository(amphibiansRemoteDataSource: amphibiansRemoteDataSource)
    }
}

